torch/csrc/autograd/python_hook.h

```diff
@@ -42,8 +46,12 @@ struct PyFunctionTensorPostAccGradHooks : public PostAccumulateGradHook {
   PyFunctionTensorPostAccGradHooks(PyObject* dict);
   ~PyFunctionTensorPostAccGradHooks() override;
   void operator()(const Variable& tensor) override;
-  // fall back to the compiled_args of PostAccumulateGradHook superclass
+  void compiled_args(torch::dynamo::autograd::CompiledNodeArgs& args) override;
+  void apply_with_saved(
+      Variable& tensor,
+      torch::dynamo::autograd::SwapSavedVariables& saved);
   PyObject* dict;
+  std::vector<int> hook_ids;
```

Comment:
This state can't go here, because if we call backwards multiple times state will leak from one session into another.  It needs to go on CompiledNodeArgs.

Reply:
Oh I see - will move. I didn't understand the lifecycle here.